### DISCUSSION TITLE
Armeria With Reactive Java (JDK 11/17)

### DISCUSSION THREAD

**Discussion by dkoval3** [line/armeria#5682] [Q&A]
I am migrating code from Spring to Armeria and have a few questions

1. I need a decorator that can reactively chain and not have a serve method with return type of HttpResponse, it must be `Single<HttpResponse>` or something equivalent 
2. When I am exposing an API with return type of `Single<T>`, I am having issues while with `Flowable<T>` it works seamlessly. I had to write a ResponseConverter with a block code and perhaps I am missing something how to expose `Single<T> `
3. It seems that there currently is not a way to return `Flowable<HttpReponse>` since there's no built-in ResponseCovnerter. What would be the alternative solution?

Any sample code would be helpful. Thanks

**Comment by minwoox** [comment]
>  I am missing something how to expose Single<T>

If you add armeria-rxjava3 depdencies, `ObservableResponseConverterFunction` will be used to convert a `Single` instance.
https://github.com/line/armeria/blob/4edff55daaad24ba6f91e21bd2eabd0073a723ef/rxjava3/src/main/java/com/linecorp/armeria/server/rxjava3/ObservableResponseConverterFunction.java#L119-L123

> What would be the alternative solution?

Could you return `HttpResponse` or `CompletableFuture<HttpReponse>` instead?

> I need a decorator that can reactively chain and not have a serve method with return type of HttpResponse, it must be Single<HttpResponse> or something equivalent

I'm not sure what you really want. 😓 Could you show me an example?

**Reply by smahad786** [reply]
I think this is what Dane is thinking about. We are able to send the request back to the API caller via CompletableFuture<HttpResponse> and/or Flowable<T>. However, when we expose a Single<T> we have to add a ResponseConverter, forcing a blockingGet(). Is it possible to avoid this blockingGet(), still return a Single<T> as a return value in our Controllers, without using a ResponseConverter to HttpResponse?

public HttpResponse convertResponse(ServiceRequestContext ctx,
      ResponseHeaders headers, @Nullable Object result, HttpHeaders trailers) throws Exception {

    if (result instanceof Single) {
      return HttpResponse.of(
          HttpStatus.OK,
          Objects.requireNonNull(headers.contentType()),
          mapper.writeValueAsString(((Single) result).blockingGet()));
    }
    // To the next response converter.
    return ResponseConverterFunction.fallthrough();
  }

Also,  we were thinking of writing an external call to OKTA via a decorator and since this is an external call, we want to keep it async. As long as the conveters are simple, just checking the headers or doing some data format conversions, it is seamless. When we have to make an external call, the converter/decorator chaining  requires blocking code. Perhaps the purpose is not to make an external calls from decorator/converter, we need this clarification.

**Reply by minwoox** [reply]
> is it possible to avoid this blockingGet(),

We need to do something like this:
```java
return HttpResponse.of(single.toCompletionStage().thenApply(content -> {
    return HttpResponse.of(HttpStatus.OK,
                           Objects.requireNonNull(headers.contentType()),
                           mapper.writeValueAsString(content));
}));
```

However, you don't have to do this by yourself. All you need to do is adding this dependency:
https://mvnrepository.com/artifact/com.linecorp.armeria/armeria-rxjava3/1.28.4
then, `ObservableResponseConverterFunction` will automatically handle it. 

> doing some data format conversions, 

Do you also need the request body from the request to call the outside server?

**Reply by smahad786** [reply]
@minwoox I added the https://mvnrepository.com/artifact/com.linecorp.armeria/armeria-rxjava3/1.28.4 dependency and returning Single<T> did not work. For some reason, only when I expose Flowable<T>, it works. I did not add any additional code to implement the ResponseConverterFunction. I understand that there is a default Implementation that will handle it and I did not see that behavior. If there is a sample code in GitHub returning Single<T> as the return value in the ReST endpoint, it will be helpful. The Mono<T> works with SpringBoot, only I see some issue with Armeria and it is something  I am missing

**Reply by ikhoon** [reply]
> dependency and returning Single did not work.

Could you provide a minimal reproducible example? Some LINE internal users have been using RxJava for a long time. If there were some issues in our implementations, we would get a report from them.

**Comment by smahad786** [comment]
Thanks @minwoox I will further look into it. I also have a follow up question. Let us test this first and then come back
